compiler/damlc/lib/DA/Cli/Options.hs

```diff
@@ -231,34 +230,104 @@ allowLargeTuplesOpt = AllowLargeTuples <$>
     where
         desc = "Do not warn when tuples of size > 5 are used."
 
--- The implementation of dlintUsageOpt allows multiple uses of dlintEnabledOpt
--- and dlintDisabledOpt, with only the last one winning. As a result,
--- dlintEnabledOpt uses strOption instead of strOptionOnce
-dlintEnabledOpt :: Parser DlintUsage
-dlintEnabledOpt = DlintEnabled
-  <$> Options.Applicative.strOption
-  ( long "with-dlint"
-    <> metavar "DIR"
-    <> internal
-    <> help "Enable linting with 'dlint.yaml' directory"
-  )
-  <*> switch
-  ( long "allow-overrides"
-    <> internal
-    <> help "Allow '.dlint.yaml' configuration overrides"
-  )
-
-dlintDisabledOpt :: Parser DlintUsage
-dlintDisabledOpt = flag' DlintDisabled
-  ( long "without-dlint"
-    <> internal
-    <> help "Disable dlint"
-  )
-
-dlintUsageOpt :: Parser DlintUsage
-dlintUsageOpt = fmap (fromMaybe DlintDisabled . lastMay) $
-  many (dlintEnabledOpt <|> dlintDisabledOpt)
+dlintRulesFileParser :: Parser DlintRulesFile
+dlintRulesFileParser =
+  lastOr DefaultDlintRulesFile $
+    defaultDlintRulesFile <|> explicitDlintRulesFile
+  where
+    defaultDlintRulesFile =
+      flag' DefaultDlintRulesFile
+        ( long "lint-default-rules"
+          <> internal
+          <> help "Use the default rules file for linting"
+        )
+    explicitDlintRulesFile =
+      ExplicitDlintRulesFile <$> strOptionOnce
+        ( long "lint-rules-file"
+          <> metavar "FILE"
+          <> internal
+          <> help "Use FILE as the rules file for linting"
+        )
+
+dlintHintFilesParser :: Parser DlintHintFiles
+dlintHintFilesParser =
+  lastOr ImplicitDlintHintFile $
+    implicitDlintHintFile <|> explicitDlintHintFiles <|> clearDlintHintFiles
+  where
+    implicitDlintHintFile =
+      flag' ImplicitDlintHintFile
+        ( long "lint-implicit-hint-file"
+          <> internal
+          <> help "Use the first '.dlint.yaml' file found in the \
+                  \project directory or any parent thereof, or, failing that, \
+                  \in the home directory of the current user."
+        )
+    explicitDlintHintFiles =
+      fmap ExplicitDlintHintFiles $
+        some $ Options.Applicative.strOption
+          ( long "lint-hint-file"
+            <> metavar "FILE"
+            <> internal
+            <> help "Add FILE as a hint file for linting. Any implicit \
+                    \'.dlint.yaml' files will be ignored."
+          )
+    clearDlintHintFiles =
+      flag' (ExplicitDlintHintFiles [])
+        ( long "lint-no-hint-files"
+          <> internal
+          <> help "Use no hint files for linting. This also ignores any \
+                  \implicit '.dlint.yaml' files"
+        )
```

Comment:
Since `--lint-hint-file` above allows multiple files to be specified, it's useful to have a way to clear them